### PR TITLE
[WIP] CMake Build System

### PR DESCRIPTION
**Pull request by mbrookhart**
Very Rough first pass of the CMake build system for #307

### CODE REVIEW

File: cmake/ngraph.cmake [hunk 1]
```diff
@@ -0,0 +1,53 @@
+include(ExternalProject)
+
+add_library(ngraph_interface_lib INTERFACE)
+
+set(EXTERNAL_NGRAPH_INSTALL_PREFIX ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/ngraph)
+set(EXTERNAL_NGRAPH_SOURCE_DIR ${CMAKE_SOURCE_DIR}/3rdparty/ngraph)
```
Comment (cconvey) [inline cmake/ngraph.cmake]: Do you want users to be able to override these?  If not, you may wish to use the [`INTERNAL` or `FORCE` options](https://cmake.org/cmake/help/v3.3/command/set.html).

Also, to reduce namespace confusion, you might consider having the variable names _start_ with `NGRAPH_`, or maybe even `MXNET_NGRAPH_`.

File: cmake/ngraph.cmake [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+ExternalProject_Add(
+	ext_ngraph
+	SOURCE_DIR ${EXTERNAL_NGRAPH_SOURCE_DIR}
+
+	PREFIX 3rdparty/ngraph/
+	CMAKE_ARGS -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+			-DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+			-DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+			-DCMAKE_INSTALL_PREFIX=${EXTERNAL_NGRAPH_INSTALL_PREFIX}
```
Comment (cconvey) [inline cmake/ngraph.cmake]: Might want to set additional cmake options for ngraph.  In particular, `NGRAPH_USE_PREBUILT_LLVM`.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -25,6 +25,15 @@ if(GTEST_FOUND AND NOT MSVC)
...
   file(GLOB_RECURSE UNIT_TEST_SOURCE "cpp/*.cc" "cpp/*.h")
+  if(NOT USE_NGRAPH)
+    FOREACH(item ${UNIT_TEST_SOURCE})
+      IF(${item} MATCHES "ngraph")
+        LIST(REMOVE_ITEM UNIT_TEST_SOURCE ${item})
+      ENDIF(${item} MATCHES "ngraph")
+    ENDFOREACH(item)
+  else()
+    include_directories(${CMAKE_SOURCE_DIR}/src/ngraph)
+  endif()
```
Comment (cconvey) [inline tests/CMakeLists.txt]: I know what you're doing this, and I agree, but it makes me want to throw up a little.  :(

**Comment by sfvaroglu** [comment]
This PR will also support USE_NGRAPH_DISTRIBUTED, right?

**Comment by mbrookhart** [comment]
Before it's done, yeah :) Very WIP, I am not a Cmake wizard

**Pull request closed without merging**